core/src/main/scala/dagr/core/tasksystem/ScatterGatherFramework.scala

```diff
@@ -0,0 +1,116 @@
+  * [[ScatterTask]]) are responsible for step (2), and a gather task (see [[GatherTask]]) is responsible for step (3).
+  *
+  * The public methods are the minimum that need to be implemented to use the scatter gather framework.
+  *
+  * Override the protected methods to customize how scatter or gather tasks are generated, for example to perform
+  * recursive merging (ala merge-sort).  Otherwise, mix in traits that do this for you.
+  *
+  * Be sure to mix-in the appropriate trait in the task returned by each method.
+  *
+  * @tparam Input        the input type to the scatter gather.
+  * @tparam Intermediate the input type to each scatter.
+  * @tparam Output       the output type of the scatter gather.
+  **/
+trait ScatterGatherFramework[Input, Intermediate, Output] extends Pipeline with GatherTask[Output] {
+
+  /** The input to the scatter gather */
+  def input: Input
+
+  /** Generates a task to split the input into multiple inputs for the scatter tasks. */
+  def splitInputTask(input: Input): SplitInputTask[Input, Intermediate]
+
+  /** Generates a scatter task from an input meant for a scatter task. */
+  def scatterTask(intermediate: Intermediate): ScatterTask[Intermediate, Output]
+
+  /** Generates a gather task that gathers an ordered set of outputs */
+  def gatherTask(outputs: Iterable[Output]): GatherTask[Output]
+
+  /** Generates the scatter tasks.  No dependencies should be set between scatter tasks and either the split input or
+    * gather tasks in this method. */
+  protected def scatterTasks(intermediates: Iterable[Intermediate]): Iterable[ScatterTask[Intermediate,Output]] = {
+    intermediates.map(in => scatterTask(in))
+  }
+
+  /** Provides the final task that contains the output of the scatter gather.  In the case that there is only one scatter,
+    * this may be a [[ScatterTask]].  Otherwise, this is more often a [[GatherTask]].  In the case of a complicated
+    * gather pipeline, say to recursively merge the outputs (ala merge-sort), the gather containing the final output
+    * should be returned.
+    *
+    * The scatter tasks are provided so in the case of a gather pipeline, individual tasks can depend on the scatter
+    * tasks. For example, not all gather tasks in a recursive merge (ala merge-sort) wil depend on the scatter tasks.
+    *
+    * The gather task(s) should have their dependencies set in this method.
+    */
+  protected def finalOutputTask(inputTasks: Iterable[ScatterTask[Intermediate,Output]]): OutputTask[Output]
+}
+
+/** All tasks that split the input into inputs for the scatter tasks should extend this trait.
+  *
+  * Provides a method that returns the inputs to the scatter tasks.  The method
+  * should only be called *after* this task has been run.  This is typically ensured by
+  * using a [[Callback]]. */
+trait SplitInputTask[Input,Intermediate] extends Task {
+  protected var _inputs: Option[Iterable[Intermediate]] = None
+  final def inputs: Iterable[Intermediate] = _inputs.get
```

Comment:
I'd be wary of using function names like `input`, `inputs`, `output` etc. in these traits, as that's going to make it really hard to mix them into task classes that have `val`s with those names.

I think the ideal usage of these traits will be something like:
```scala
new IntervalListTools(....) with SplitInputTask {
  def scatteredInputs = ...
}
```